presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,759 @@
+                    sourceTypes,
+                    pagePreprocessor,
+                    partitionFunction,
+                    partitionChannels,
+                    partitionConstants,
+                    replicatesAnyRow,
+                    nullChannel,
+                    outputBuffer,
+                    serdeFactory,
+                    maxMemory);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new OptimizedPartitionedOutputOperatorFactory(
+                    operatorId,
+                    planNodeId,
+                    sourceTypes,
+                    pagePreprocessor,
+                    partitionFunction,
+                    partitionChannels,
+                    partitionConstants,
+                    replicatesAnyRow,
+                    nullChannel,
+                    outputBuffer,
+                    serdeFactory,
+                    maxMemory);
+        }
+    }
+
+    private static class PagePartitioner
+    {
+        private static final int LONG_DECIMAL_SERIALIZED_BYTES = 17;
+        private static final int SHORT_DECIMAL_SERIALIZED_BYTES = 9;
+        private static final Map<String, Integer> FIXED_WIDTH_TYPE_SERIALIZED_BYTES = ImmutableMap.<String, Integer>builder()
+                .put(ColorType.NAME, 5)
+                .put(StandardTypes.BIGINT, 9)
+                .put(StandardTypes.BING_TILE, 9)
+                .put(StandardTypes.BOOLEAN, 2)
+                .put(StandardTypes.INTEGER, 5)
+                .put(StandardTypes.DATE, 5)
+                .put(StandardTypes.DOUBLE, 9)
+                .put(StandardTypes.INTERVAL_DAY_TO_SECOND, 9)
+                .put(StandardTypes.INTERVAL_YEAR_TO_MONTH, 5)
+                .put(StandardTypes.IPADDRESS, 17)
+                .put(StandardTypes.REAL, 5)
+                .put(StandardTypes.SMALLINT, 3)
+                .put(StandardTypes.TIME, 9)
+                .put(StandardTypes.TIME_WITH_TIME_ZONE, 9)
+                .put(StandardTypes.TIMESTAMP, 9)
+                .put(StandardTypes.TIMESTAMP_WITH_TIME_ZONE, 9)
+                .put(StandardTypes.TINYINT, 2)
+                .put(UnknownType.NAME, 2)
+                .build();
+
+        private final OutputBuffer outputBuffer;
+        private final PartitionFunction partitionFunction;
+        private final List<Integer> partitionChannels;
+        private final List<Optional<Block>> partitionConstants;
+        private final PagesSerde serde;
+        private final boolean replicatesAnyRow;
+        private final OptionalInt nullChannel; // when present, send the position to every partition if this channel is null.
+        private final AtomicLong rowsAdded = new AtomicLong();
+        private final AtomicLong pagesAdded = new AtomicLong();
+
+        private final BlockFlattener flattener = new BlockFlattener(new SimpleArrayAllocator());
```

Comment:
@arhimondr After discussion with @mbasmanova, we actually didn't make SimpleArrayAllocator hidden inside the BlockFlattener. We still create a new SimpleArrayAllocator and pass it in to BlockFlattener. But we do count for the memory of the SimpleArrayAllocator. Because the same SimpleArrayAllocator is not used by multiple BlockFlatteners or operators, the counting should be accurate  since they're 1-1 mappings. The code change is in 95416b2b8c "Add getRetainedSizeInBytes to BlockFlattener"